cadc-download-manager-server/src/main/java/ca/nrc/cadc/dlm/server/ScriptGenerator.java

```diff
@@ -0,0 +1,154 @@
+ */
+class ScriptGenerator {
+    private static final String SCRIPT_HEADER = "#!/bin/bash";
+    private static final String NEWLINE = "\n";
+    private static final String TOKEN_EXPIRY_WARNING =
+            "Warning!  The token provided in the cURL commands will expire on %s.";
+    private static final String TOKEN_SCRIPT_VARIABLE = "TOKEN=\"%s\"";
+    private static final String ANON_CURL_COMMAND_PREFIX = "curl --location --remote-name --remote-header-name --progress-bar ";
+    private static final String AUTH_CURL_COMMAND_PREFIX =
+            ScriptGenerator.ANON_CURL_COMMAND_PREFIX + "-H \"authorization: bearer ${TOKEN}\" ";
+    private static final String ECHO = "echo ";
+    private static final String TOKEN_EXPIRY_WARNING_COMMAND = ScriptGenerator.ECHO + "\""
+                                                               + ScriptGenerator.TOKEN_EXPIRY_WARNING + "\"";
+
+    private final Iterator<DownloadDescriptor> downloadDescriptors;
+    private final String authToken;
+    private final Date expiryDate;
+
+
+    ScriptGenerator(final Iterator<DownloadDescriptor> downloadDescriptors) {
+        this(downloadDescriptors, null, null);
+    }
+
+    ScriptGenerator(final Iterator<DownloadDescriptor> downloadDescriptors, final String authToken,
+                           final Date expiryDate) {
+        this.downloadDescriptors = downloadDescriptors;
+        this.authToken = authToken;
+        this.expiryDate = expiryDate;
+
+        // If the auth token was provided, the expiry date should be provided as well.
+        if (this.expiryDate == null && StringUtil.hasLength(this.authToken)) {
+            throw new IllegalArgumentException("The Expiry Date is required when the token is supplied.");
+        } else if (this.downloadDescriptors == null || !this.downloadDescriptors.hasNext()) {
+            throw new IllegalArgumentException("The DownloadDescriptor iterator is required.");
+        }
+    }
+
+
+    void generate(final Writer writer) throws IOException {
+        writer.write(ScriptGenerator.SCRIPT_HEADER);
+        writer.write(ScriptGenerator.NEWLINE);
+
+        final String curlCommandPrefix;
+        if (StringUtil.hasLength(this.authToken)) {
+            writer.write(ScriptGenerator.NEWLINE);
+            curlCommandPrefix = ScriptGenerator.AUTH_CURL_COMMAND_PREFIX;
+            writer.write(String.format(TOKEN_EXPIRY_WARNING_COMMAND,
+                                       DateUtil.getDateFormat(DateUtil.ISO8601_DATE_FORMAT_LOCAL, DateUtil.UTC)
+                                               .format(this.expiryDate)));
+            writer.write(ScriptGenerator.NEWLINE);
+            writer.write(ScriptGenerator.NEWLINE);
+            writer.write(String.format(ScriptGenerator.TOKEN_SCRIPT_VARIABLE, this.authToken));
+            writer.write(ScriptGenerator.NEWLINE);
+        } else {
+            curlCommandPrefix = ScriptGenerator.ANON_CURL_COMMAND_PREFIX;
+        }
+
```

Comment:
The script might be more readable if you put the URLs into a bash array and then loop : the curl command in use would be printed once instead of many times. 

You would have to iterate to separate the URLs and uri+error and then iterate again to write the bash arrays, and need a bash `for` loop for each, and decide to put errors first or last...

Reply:
Google says that Bash arrays have no maximum size, so it sounds plausible.  Aren't individual `cURL` commands more readable though?  Some subset could be copied and pasted easily and still work as-is.

Reply:
This has been reworked.  See the template script and the test expectations.